js/patternkit.jsoneditor.es6.js

```diff
@@ -0,0 +1,191 @@
+        window.patternkitEditor.getEditor(response.selector).setValue(response.value);
+      };
+      let saveSchema = function () {
+        $('#schema_instance_config').val(JSON.stringify(window.patternkitEditor.getValue()));
+        if (window.M) {
+          window.M.updateTextFields();
+        }
+      };
+      let $target = $('#editor-shadow-injection-target', context);
+      $target.once('patternkit-editor').each(function () {
+        let shadow = this.attachShadow({mode: 'open'});
+        let theme_js = settings.patternkitEditor.themeJS;
+        if (typeof theme_js === 'string') {
+          theme_js = [theme_js];
+        }
+        for (let i = 0; theme_js && i < theme_js.length; i++) {
+          let script_element = document.createElement('script');
+          script_element.type = "text/javascript";
+          script_element.src = theme_js[i];
+          document.getElementsByTagName('head')[0].appendChild(script_element);
+        }
+        let editor_dom = '';
+        if (settings.patternkitEditor.themeStylesheet) {
+          editor_dom = '<link rel="stylesheet" id="theme_stylesheet" href="' + settings.patternkitEditor.themeStylesheet + '">';
+        }
+        // @todo Re-eval with this shadow dom webfont bug:
+        // https://bugs.chromium.org/p/chromium/issues/detail?id=336876
+        if (settings.patternkitEditor.iconStylesheet) {
+          let icons_element = document.createElement('link');
+          icons_element.rel = "stylesheet";
+          icons_element.id = "icon_stylesheet";
+          icons_element.href = settings.patternkitEditor.iconStylesheet;
+          document.getElementsByTagName('head')[0].appendChild(icons_element);
+          editor_dom += '<link rel="stylesheet" id="icon_stylesheet" href="' + settings.patternkitEditor.iconStylesheet + '">';
+        }
+        editor_dom += '<div id="editor_holder"></div>';
+        shadow.innerHTML += editor_dom;
+        let data = {
+          schema: JSON.parse(settings.patternkitEditor.schemaJson),
+          starting: JSON.parse(settings.patternkitEditor.startingJson)
+        };
+        JSONEditor.defaults.options.theme = settings.patternkitEditor.theme;
+        JSONEditor.defaults.options.iconlib = settings.patternkitEditor.icons;
+        JSONEditor.defaults.options.keep_oneof_values = false;
+        JSONEditor.defaults.options.disable_edit_json = true;
+        JSONEditor.defaults.options.disable_collapse = false;
+        JSONEditor.defaults.options.collapse = false;
+        JSONEditor.defaults.options.ajax = true;
+        JSONEditor.defaults.editors.drupal_image = DrupalImageEditor;
+        JSONEditor.defaults.resolvers.unshift(function (schema) {
+          if (schema.type === 'string' && schema.format === 'image') {
+            return 'drupal_image';
+          }
+        });
+        // Override how references are resolved.
+        JSONEditor.prototype._loadExternalRefs = function (schema, callback) {
```

Comment:
why is this needed?

Reply:
This allows us to load refs using URL requests to the patternkit API, which resolves to a file ref.

The API could also hit a REST endpoint URL that returns the schema etc etc.

This allows us to use OOTB JSON Schema as well as namespace-based refs without needing to recursively load all schemas into the editor.